src/hotspot/share/utilities/zipLibrary.cpp

```diff
@@ -0,0 +1,186 @@
+typedef void(*ZIP_Close_t)(jzfile* zip);
+typedef jzentry* (*ZIP_FindEntry_t)(jzfile* zip, const char* name, jint* sizeP, jint* nameLen);
+typedef jboolean(*ZIP_ReadEntry_t)(jzfile* zip, jzentry* entry, unsigned char* buf, char* namebuf);
+typedef jint(*ZIP_CRC32_t)(jint crc, const jbyte* buf, jint len);
+typedef const char* (*ZIP_GZip_InitParams_t)(size_t, size_t*, size_t*, int);
+typedef size_t(*ZIP_GZip_Fully_t)(char*, size_t, char*, size_t, char*, size_t, int, char*, char const**);
+
+static ZIP_Open_t ZIP_Open = nullptr;
+static ZIP_Close_t ZIP_Close = nullptr;
+static ZIP_FindEntry_t ZIP_FindEntry = nullptr;
+static ZIP_ReadEntry_t ZIP_ReadEntry = nullptr;
+static ZIP_CRC32_t ZIP_CRC32 = nullptr;
+static ZIP_GZip_InitParams_t ZIP_GZip_InitParams = nullptr;
+static ZIP_GZip_Fully_t ZIP_GZip_Fully = nullptr;
+
+static void* _zip_handle = nullptr;
+static bool _loaded = false;
+
+static inline bool is_loaded() {
+  return Atomic::load_acquire(&_loaded);
+}
+
+static inline bool not_loaded() {
+  return !is_loaded();
+}
+
+static void* dll_lookup(const char* name, const char* path) {
+  assert(_zip_handle != nullptr, "invariant");
+  void* func = os::dll_lookup(_zip_handle, name);
+  if (func == nullptr) {
+    char msg[256] = "";
+    jio_snprintf(&msg[0], sizeof msg, "Could not resolve \"%s\"", name);
+    vm_exit_during_initialization(&msg[0], path);
+  }
+  return func;
+}
+
+static void store_function_pointers(const char* path) {
+  assert(_zip_handle != nullptr, "invariant");
+  ZIP_Open = CAST_TO_FN_PTR(ZIP_Open_t, dll_lookup("ZIP_Open", path));
+  ZIP_Close = CAST_TO_FN_PTR(ZIP_Close_t, dll_lookup("ZIP_Close", path));
+  ZIP_FindEntry = CAST_TO_FN_PTR(ZIP_FindEntry_t, dll_lookup("ZIP_FindEntry", path));
+  ZIP_ReadEntry = CAST_TO_FN_PTR(ZIP_ReadEntry_t, dll_lookup("ZIP_ReadEntry", path));
+  ZIP_CRC32 = CAST_TO_FN_PTR(ZIP_CRC32_t, dll_lookup("ZIP_CRC32", path));
+  ZIP_GZip_InitParams = CAST_TO_FN_PTR(ZIP_GZip_InitParams_t, dll_lookup("ZIP_GZip_InitParams", path));
+  ZIP_GZip_Fully = CAST_TO_FN_PTR(ZIP_GZip_Fully_t, dll_lookup("ZIP_GZip_Fully", path));
+}
+
+static void load_zip_library() {
+  assert(!is_loaded(), "should not load zip library twice");
+  char path[JVM_MAXPATHLEN];
+  if (os::dll_locate_lib(&path[0], sizeof path, Arguments::get_dll_dir(), "zip")) {
+    char ebuf[1024];
+    _zip_handle = os::dll_load(&path[0], &ebuf[0], sizeof ebuf);
+  }
+  if (_zip_handle == nullptr) {
+    vm_exit_during_initialization("Unable to load zip library", &path[0]);
+  }
+  store_function_pointers(&path[0]);
+  Atomic::release_store(&_loaded, true);
+  assert(is_loaded(), "invariant");
+}
+
+//
+// Helper mutex class that also ensures that java threads
+// are in _thread_in_native when loading the zip library.
+//
+class ZipLibraryLoaderLock : public StackObj {
+ private:
+  static Semaphore _lock;
+  JavaThread* _jt;
+ public:
+   ZipLibraryLoaderLock() : _jt(nullptr) {
+    Thread* thread = Thread::current_or_null_safe();
```

Comment:
Why do you need the "safe" version here? can this be called from a signal handling context?

Reply:
I needed the safe version here because it can be called early enough in the VM bootstrap where the current thread is not yet initialized. Will check again where this comes from during bootstrap.

Reply:
Updated to Thread::current_or_null().